CPS-???? | Governance Security (#490)
cardano-foundation/CIPs: CIP-????/README.md

---

> CIP: ???

Author: Yes @Ryun1 I just read CPS and I think you are correct. I'm going to switch to that format and continue from there. Once I move to that format can I get your input on the Problem and Use Cases sections of a CPS?

What I am thinking is that I can put the current 25 different line items entirely in the Problem section, or the other alternative is to break it up into 25 different CPS with each having its own CPS. I prefer to keep it simple as one CPS though.

Title: Cardano Governance Security
Category: Ledger
Status: Proposed
Authors:
    - Richard McCracken <rickymac68@icloud.com>
    - Andrew Westberg <andrewwestberg@gmail.com>
Implementors: N/A
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/490
Created: 2023-03-27
License: CC-BY-4.0
---

> # Abstract

Reviewer: ```suggestion
## Abstract
```
Each of these document sections needs to be demoted to H2 (as it also says in the CIP and CPS templates) so an H1 title can be prepended as a document title from the information that's in the header.

Author: Thanks @rphair I'll get the other sections as well.

> As Cardano community led governance comes online, security of the governance mechanism must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constittional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidelines also contain as to how wallet developers, dApp developers, and infrastructure developers in general to help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security.

Reviewer 2: ```suggestion
As Cardano community led governance comes online, security of the governance mechanism must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constitutional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidelines also contain as to how wallet developers, dApp developers, and infrastructure developers in general to help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security.

```

> # Abstract
> 
> As Cardano community led governance comes online, security of the governance mechanism must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constittional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidelines also contain as to how wallet developers, dApp developers, and infrastructure developers in general to help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security.
> 

Reviewer 3: ```suggestion
# Abstract

As Cardano community led governance comes online, security of the governance mechanism must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constittional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidelines also pertain to how wallet developers, dApp developers, and infrastructure developers in general help build situational awareness so that, as development progresses in these areas, consideration in design is given to governance security.
```

After making your changes, you can add additional context and explanation of your changes and why you view them as necessary down here.

Reviewer 4: I would have preferred something simpler, an easy read like: "The guidelines also cover how wallet developers, dApp developers, and infrastructure developers in general help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security"

Reviewer 2: Constittional = Constitutional

# Motivation

> This CIP intent is to promote awareness of governance security vulnerabilities and possible mitigation techniques in the Cardano ecosystem. The goal is to provide a comprehensive guide describing all known governance security threats and possible mitigiations techniques, but not to address specific software vulnerabilities. The governance security need is present regardless of the final form of CIP-1694 or any other governance document to minimize financial damage and nefarious activities that may occur when a system of voting, elections, and on-chain approval of automated actions are present.

Reviewer 5: ```suggestion
This CIP intent is to promote awareness of governance security vulnerabilities and possible mitigation techniques in the Cardano ecosystem. The goal is to provide a comprehensive guide describing all known governance security threats and possible mitigiations techniques, but not to address specific software vulnerabilities. The governance security need is present regardless of the final form of CIP-1694 or any other governance document to minimize malicious activities that may occur when a system of voting, elections, and/or on-chain approval of automated actions are present.
```

> # Specification

Reviewer: ```suggestion
## Goals
```
(as stated above)

> Governance on Cardano in this CIP is defined as on chain decision making that is automatically executed when some threshold of approval by human actors is met. Each threat or vulnerability will be described as a "threat", "description", and the suggested mitigiation technique will be described as a "recommendation" for each type of threat identified in this document. All known threats and mitigations will be made available to public as broadly as possible. The more voters, developers, and governance actors are aware of threats the more likely they are to be identified and mitigated before damage occurs.

Reviewer 2: ```suggestion
Governance on Cardano in this CIP is defined as on-chain decision making that is automatically executed when some threshold of approval by human actors is met. Each threat or vulnerability will be described as a "threat", "description", and the suggested mitigiation technique will be described as a "recommendation" for each type of threat identified in this document. All known threats and mitigations will be made available to public as broadly as possible. The more voters, developers, and governance actors are aware of threats the more likely they are to be identified and mitigated before damage occurs.

```

> On rare occation a threat may be identified that requires responsible reporting procedures in order to prevent exposing an active vulnerability to bad actors who may aim to exploit the vulnerability. In the event responsible reporting is required, notify the appropriate group or entity with the responsiblility and resources to mitigate such threat in a confidential manner.

Reviewer 6: Line 29 rtypos-
```suggestion
On rare occasion a threat may be identified that requires responsible reporting procedures in order to prevent exposing an active vulnerability to bad actors who may aim to exploit the vulnerability. In the event responsible reporting is required, notify the appropriate group or entity with the responsibility and resources to mitigate such threat in a confidential manner.
```

### #2: Emergency group to support the disaster recover plan.
  -attacks on persons if known
  -we don’t want attackers know who that group is?

Recommendation:

### #3: Emergency change that needs to be done quickly.
  -a nuclear option
  -change requires “just do it” 
  -a security that fix must be implemented but CANNOT be communicated publicly

Recommendation:

### #4: Operational Security (OPSEC) of the committee.
  -physical security
  -digital security

Recommendation:

### #5: Committee keys lost or stolen or committee keys sold to a bad actor.

Recommendation:

### #6: DRep keys lost or stolen or DRep keys sold to a bad actor.

Recommendation:

### #7: Lack of governance for an extended period of time.
  -Uncontrolled hard forks by SPOs
  -Risk of no confidence

Recommendation:

### #8: Lack of trust in voter tooling.

Recommendation:

### #9: Trust in the wallets.
  -wallet voting attack vector
  -hardware wallet support

Recommendation:

### #10: Corruption - in any form, any number of actors.
  -Code of Conduct mitigation

>   -on chain mitigation

Reviewer 2: ```suggestion
  -on-chain mitigation

```

Recommendation:

### #15: Persistence - Removal of proposals by attackers.

Recommendation:

### #16: Risk of Proposal overload - Risk of DDOS.

Recommendation:

### #17: Excessive actions.
  -Drawing too much money
  -Changing parameters that break the chain (block size=0 for example)
  -overlapping actions and duplicate actions

Recommendation:

### #18: Lack of technical understanding by DReps.
Education of correct procedures for DReps

Recommendation:

### #19: Sybil behavior - one person acting as multiple DReps.

Recommendation:

### #20: Overpower Risk - One DRep or entity with too much power.

Recommendation:

### #21: Superstar attack - one “superstar” person that everybody delegates to becomes a dictator with little skin in the game.
(Example: Elon Musk as a DRep)

### #22: High chain load impacting governance.

Recommendation:

### #23: Unknown external regulation risks - affecting a specific DRep by country or affecting governance as a whole.

Recommendation:

### #24: Blocking a Constitution Committee revote after a vote of “no confidence”.

Recommendation:

### #25: Ensuring the legitimacy of side chains involved in governance.

Recommendation:

> # Rational

Reviewer 6: A Rationale

> The threats and recommendations listed here are the result of consultation among cyber security experts, financial experts, game theory experts, and Voltaire workshop attendees which included developoers, stake pool operators, representatives from IOG, Cardano Foundation and Emurgo, memebers of Project Catalyst, and Cardano community members at large.

Reviewer 7: ```suggestion
The threats and recommendations listed here are the result of consultation among cyber security experts, financial experts, game theory experts, and Voltaire workshop attendees which included developers, stake pool operators, representatives from IOG, Cardano Foundation and Emurgo, members of Project Catalyst, and Cardano community members at large.
```

## Goals

Governance on Cardano in this CIP is defined as on-chain decision making that is automatically executed when some threshold of approval by human actors is met. Each threat or vulnerability will be described as a "threat", "description", and the suggested mitigiation technique will be described as a "recommendation" for each type of threat identified in this document. All known threats and mitigations will be made available to public as broadly as possible. The more voters, developers, and governance actors are aware of threats the more likely they are to be identified and mitigated before damage occurs.

On rare occasion a threat may be identified that requires responsible reporting procedures in order to prevent exposing an active vulnerability to bad actors who may aim to exploit the vulnerability. In the event responsible reporting is required, notify the appropriate group or entity with the responsibility and resources to mitigate such threat in a confidential manner.

Assumption: There are no known software back doors. Unknown software vulnerabilities may be discovered.

### #1: Unknown software vulnerability affecting governance is discovered.

> Recommendation: The person discovering the vulnerability should document the circumstances and details in which the vulnerability is present and how to reproduce the vulnerability. Then report the vulnerability in a confidential manner using a secure method to the developer team with the resources to correct the vulnerability.

Reviewer 4: The last part "_with the resources to correct the vulnerability_" may not be known to the person that highlights the vulnerability. It can be optional.

> Recommendation: The person discovering the vulnerability should document the circumstances and details in which the vulnerability is present and how to reproduce the vulnerability. Then report the vulnerability in a confidential manner using a secure method to the developer team with the resources to correct the vulnerability.

Reviewer 4: The person discovering the vulnerability should document the circumstances and details in which the vulnerability is present and how to reproduce the vulnerability. The vulnerability should then be reported in a confidential manner using a secure method to the developer team. If the fix is known, the person can propose approaches/resources to correct the vulnerability

> Recommendation: The person discovering the vulnerability should document the circumstances and details in which the vulnerability is present and how to reproduce the vulnerability. Then report the vulnerability in a confidential manner using a secure method to the developer team with the resources to correct the vulnerability.

Reviewer 8: A tiered bug bounty rewards program can be employed to recompense those who articulate meaningful corrections required to reverse a potential vulnerability. A treasury may need to be formed to fund this specific activity, releasing those proposed rewards by a committee vote.

### #2: Emergency group to support the disaster recovery plan.

>   -attacks on persons if known

Reviewer 4: I think "exposes the person(s) if known" is better. It is efficient to focus our resources in forward thinking than retribution. If at the end, some people want to take that extra step, anyone can certainly do so.

>   -we don’t want attackers know who that group is?

Reviewer 8: "we don’t want attackers **to** know who that group is?"

Recommendation:

### #3: Emergency change that needs to be done quickly.
  -a nuclear option
  -change requires “just do it” 
  -a security that fix must be implemented but CANNOT be communicated publicly

Recommendation:

### #4: Operational Security (OPSEC) of the committee.
  -physical security
  -digital security

Recommendation:

### #5: Committee keys lost or stolen or committee keys sold to a bad actor.

> Recommendation:

Reviewer 4: - Keys should be held in a decentralized manner so that one person can't have it all
- When a committee member loses portion of their key, new key can be generated and all members get new portions as a result

### #6: DRep keys lost or stolen or DRep keys sold to a bad actor.

Recommendation:

> 

Reviewer 8: - Lost/Stolen : DRep must be able to sign with the same wallet/stake key to authenticate for burn/mint of previous key

- Sold to bad actor : Loss of collateral / Keys destroyed by committee vote

### #7: Lack of governance for an extended period of time.
  -Uncontrolled hard forks by SPOs
  -Risk of no confidence

> 

Reviewer 4: Please explain what the cause of this lack of action is. That would give more context for better suggestions

Recommendation:

### #8: Lack of trust in voter tooling.

Recommendation:

### #9: Trust in the wallets.
  -wallet voting attack vector
  -hardware wallet support

Recommendation:

### #10: Corruption - in any form, any number of actors.
  -Code of Conduct mitigation
  -on-chain mitigation

Recommendation:

### #11: Collusion - the bad version of collaboration to do something illegal or undesirable.

Recommendation:

> 

Reviewer 8: If this is collusion on the part of any committees members or DReps, it should result in loss of collateral by committee vote. Evidence needs to be presented in a reasonable and articulable way to a body (i.e. Oversight & Accountability Committee), with specific constitutional citations that explains the behavior and/or actions. Voting on such an action must be constitutionally based, not emotionally based.

### #12: Vote buying - Ada kickbacks.

Recommendation:

### #13: IGCO - initial governance coin offering
  -offering voters a token of unknown value by a representative to acquire delegators
  -including tokens other than Ada

Recommendation:

### #14: Powerful Voter Collusion
  -demotivates smaller bag holders
  -cause a loss of participation
  -too many voters abstain
  -minority controls the outcome
  -can push through any proposal

> Recommendation:

Reviewer 4: **Explore** quadratic voting or other normalizing strategies to help mitigate minority rule.

### #15: Persistence - Removal of proposals by attackers.

Recommendation:

> 

Reviewer 8: I know I continue to suggest collateral. I'm going to do it here, too. If an individual or group of individuals want to remove a proposal, collateral should be required, just as submitting the proposal should. If the removal request is found by committee to be nefarious and/or without merit, a loss of collateral should occur.

### #16: Risk of Proposal overload - Risk of DDOS.

Recommendation:

> 

Reviewer 9: - Requiring a minimum amount of ada to be staked by a stake address to create a proposal
- Allowing each stake key to only have one proposal at a time
- A "cool down" period where a proposer cannot submit a new proposal for XX amount of time after the conclusion of their previous proposal  for a stake address after their last proposal's vote was concluded
- A "warm up" period that prevents newly delegated ada from proposing for a certain period of time (preventing bad actors from circumnavigating the above)

### #17: Excessive actions.
  -Drawing too much money
  -Changing parameters that break the chain (block size=0 for example)
  -overlapping actions and duplicate actions

Recommendation:

> 

Reviewer 8: - Intentional attempts to submit parameter change proposals that are deemed detrimental to the health of the network should result in a loss of collateral. As mentioned for line 118, collateral should be required to submit proposals.
- overlapping/duplicate actions should be mostly curbed due to collateral requirement. If 1/1 duplicate is successfully proposed, the enforcement of a penalty from collateral as a percentage might be appropriate.

### #18: Lack of technical understanding by DReps.
Education of correct procedures for DReps

Recommendation:

> 

Reviewer 9: - The creation of peer-reviewed content to explain a technology in terms anyone can understand, and to independently summarize and verify the pros and cons of a specific technical change. (maybe funded by a treasury to improve impartiality?)

> 

Reviewer 8: Curriculum framed from the contents of the constitution and the approved governance CIP (CIP-1694 or otherwise)

### #19: Sybil behavior - one person acting as multiple DReps.

Recommendation:

> 

Reviewer 8: - Loss of collateral and expungement of registration. DReps should also be required to submit collateral with their registration. Not dissimilar from a minimum pledge with a stake pool.

### #20: Overpower Risk - One DRep or entity with too much power.

Recommendation:

### #21: Superstar attack - one “superstar” person that everybody delegates to becomes a dictator with little skin in the game.
(Example: Elon Musk as a DRep)

> 

Reviewer 9: Recommendation:

> 

Reviewer 8: - Collateral exceptions or tiers - A potential process during DRep registration where an approval/finalization process occurs by committee. A tiered collateral model might be appropriate, levied based on certain criteria (i.e. social influence, history of nefarious behavior)

### #22: High chain load impacting governance.

Recommendation:

> 

Reviewer 8: - run all governance actions on a sidechain
- gAda (Governance Ada) at 1:1 Mainnet Ada
- Snapshot based (every epoch)

### #23: Unknown external regulation risks - affecting a specific DRep by country or affecting governance as a whole.

Recommendation:

> 

Reviewer 8: - Form a Regulation Preparedness Committee
- All issues of importance can be brought to the floor of the committee for evaluation, discussion, recommendation and/or action

### #24: Blocking a Constitutional Committee revote after a vote of “no confidence”.

Recommendation:

> 

Reviewer 8: - Proposal should be returned to the DReps with documentation that outlines why a vote of "no confidence" was given, giving the DRep body an opportunity to submit reasons and potential solutions to the proposer(s)
- or a direct revote by the DRep body requiring a strong percentage (3/4?) of votes to the affirmative in order to successfully override the "no confidence" vote.

### #25: Ensuring the legitimacy of side chains involved in governance.

Recommendation:

> 

Reviewer 8: - gNodes (Governance Nodes) implemented by process of election
- limit the number of nodes to (100-200)
- disallow DReps and other committee members from operating gNodes
- conflicts of interest should be outlined (i.e. spouses of gNode operators cannot be DReps and vice versa)
- risk of collateral/pledge loss for circumventing outlined accepted practices
- outline requirements for operating a gNode

---
CIP: ???
Title: Cardano Governance Security
Category: Ledger
Status: Proposed
Authors:
    - Richard McCracken <rickymac68@icloud.com>
    - Andrew Westberg <andrewwestberg@gmail.com>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/490
Created: 2023-03-27
License: CC-BY-4.0
---

# Abstract

> As Cardano community led governance comes online, security of the governance mechanisme must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constittional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidlines also contain as to how wallet developers, dApp developers, and infrastructure developers in general to help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security.

Reviewer 3: ```suggestion
As Cardano community led governance comes online, security of the governance mechanism must be addressed in a manner so that all participants are aware of the security weaknesses and how each weakness can be addressed. The guidelines contained in the CIP are relevant to Constittional Committe members, Delegation Representatives, Stake Pool Operators, voters, and Ada holders at large. The guidelines also contain as to how wallet developers, dApp developers, and infrastructure developers in general to help build situational awareness so that as development progresses in these areas, consideration in design is given to governance security.
```

# Specification

Governance on Cardano in this CIP is defined as on chain decision making that is automatically executed when some threshold of approval by human actors is met. Each threat or vulnerability will be described as a "threat", "description", and the suggested mitigiation technique will be described as a "recommendation" for each type of threat identified in this document. All known threats and mitigations will be made available to public as broadly as possible. The more voters, developers, and governance actors are aware of threats the more likely they are to be identified and mitigated before damage occurs.

On rare occation a threat may be identified that requires responsible reporting procedures in order to prevent exposing an active vulnerability to bad actors who may aim to exploit the vulnerability. In the event responsible reporting is required, notify the appropriate group or entity with the responsiblility and resources to mitigate such threat in a confidential manner.

Assumption: there are no known software back doors.

> 

Author: ```suggestion
1. Possible unknown software vulnerabilities.
```
